scipy/stats/_sensitivity_analysis.py

```diff
@@ -0,0 +1,634 @@
+        + 0.1 * (x[2]**4) * np.sin(x[0])
+    )
+    return f_eval
+
+
+def sample_A_B(
+    n: IntNumber,
+    dists: List[PPFDist],
+    random_state: SeedType = None
+) -> np.ndarray:
+    """Sample two matrices A and B.
+
+    Uses a Sobol' sequence with 2`d` columns to have 2 uncorrelated matrices.
+    This is more efficient than using 2 random draw of Sobol'.
+    See sec. 5 from [1]_.
+
+    Output shape is (d, n).
+
+    References
+    ----------
+    .. [1] Saltelli, A., P. Annoni, I. Azzini, F. Campolongo, M. Ratto, and
+       S. Tarantola. "Variance based sensitivity analysis of model
+       output. Design and estimator for the total sensitivity index."
+       Computer Physics Communications, 181(2):259-270,
+       :doi:`10.1016/j.cpc.2009.09.018`, 2010.
+    """
+    d = len(dists)
+    A_B = qmc.Sobol(d=2*d, seed=random_state, bits=64).random(n).T
+    A_B = A_B.reshape(2, d, -1)
+    try:
+        for d_, dist in enumerate(dists):
+            A_B[:, d_] = dist.ppf(A_B[:, d_])
+    except AttributeError as exc:
+        message = "Each distribution in `dists` must have method `ppf`."
+        raise ValueError(message) from exc
+    return A_B
+
+
+def sample_AB(A: np.ndarray, B: np.ndarray) -> np.ndarray:
+    """AB matrix.
+
+    AB: columns of B into A. Shape (d, n*d). e.g in 2d
+    Take A and replace 1st column with 1st column of B. You have (d, n)
+    Then A and replace 2nd column with 2nd column of B. You have (d, n)
+    Concatenate to get AB. Which means AB is (d, 2*n).
+    """
+    d, n = A.shape
+    AB = np.tile(A, (d, 1, 1))
+    i = np.arange(d)
+    AB[i, i] = B[i]
+    AB = np.concatenate(AB, axis=1)
+    return AB
+
+
+def saltelli_2010(
+    f_A: np.ndarray, f_B: np.ndarray, f_AB: np.ndarray
+) -> Tuple[np.ndarray, np.ndarray]:
+    r"""Saltelli2010 formulation.
+
+    .. math::
+
+        S_i = \frac{1}{N} \sum_{j=1}^N
+        f(\mathbf{B})_j (f(\mathbf{AB}^{(i)})_j - f(\mathbf{A})_j)
+
+    .. math::
+
+        S_{T_i} = \frac{1}{N} \sum_{j=1}^N
+        (f(\mathbf{A})_j - f(\mathbf{AB}^{(i)})_j)^2
+
+    Parameters
+    ----------
+    f_A, f_B, f_AB : array_like (s, n)
+        Function evaluations ``n`` being the number of evaluations and ``s``
+        a vector output.
+
+    Returns
+    -------
+    s, st : array_like (s, d)
+        First order and total order Sobol' indices.
+
+    References
+    ----------
+    .. [1] Saltelli, A., P. Annoni, I. Azzini, F. Campolongo, M. Ratto, and
+       S. Tarantola. "Variance based sensitivity analysis of model
+       output. Design and estimator for the total sensitivity index."
+       Computer Physics Communications, 181(2):259-270,
+       :doi:`10.1016/j.cpc.2009.09.018`, 2010.
+    """
+    s_, n = f_A.shape
+    f_AB = f_AB.reshape((-1, s_, n))
+
+    # Empirical variance calculated using output from A and B which are
+    # independent. Output of AB is not independent and cannot be used
+    var = np.var(np.concatenate([f_A, f_B], axis=1), axis=1)
```

Comment:
I believe this hits the issue of constant outputs within a given subset leading to non-zero variance values

See: https://github.com/numpy/numpy/issues/9631

In SALib, we have an explicit check for this case

https://github.com/SALib/SALib/blob/b824d129ecbc1edc8827206eb6d8b6f025ed4f8a/src/SALib/analyze/sobol.py#L198

Reply:
@ConnectedSystems I could add a check outside of this function so that t would be done for any indices function user would use (and us if we add other methods at some point).